source/yats/include/yats/scheduler.h

```diff
@@ -1,90 +1,219 @@
+        while(true)
+        {
+            abstract_task_container* task = nullptr;
             {
-                auto runnable = std::find_if(to_run.begin(), to_run.end(), [](abstract_task_container* task) {
-                    return task->can_run();
+                std::unique_lock<std::mutex> lock(m_mutex_main_thread);
+                m_wait_for_main_thread_task.wait(lock, [this]()
+                {
+                    return m_tasks_processed <= m_tasks.size();
                 });
 
-                if (runnable == to_run.end())
+                if (!m_tasks_to_process_in_main_thread.empty())
                 {
-                    return;
+                    task = m_tasks_to_process_in_main_thread.front();
+                    m_tasks_to_process_in_main_thread.pop();
                 }
-                auto task = *runnable;
-                to_run.erase(runnable);
-
-                m_threads.emplace_back([&to_run, &check_runnable, task]() {
-                    task->run();
-                    check_runnable();
-                });
-            }
-        };
 
-        check_runnable();
-        while (true)
-        {
-            std::this_thread::sleep_for(std::chrono::milliseconds(10));
-            {
-                std::lock_guard<std::mutex> lock(m_mutex);
-                if (to_run.empty())
+                if (m_tasks_processed >= m_tasks.size())
                 {
                     break;
                 }
             }
-        }
 
-        for (auto& thread : m_threads)
-        {
-            thread.join();
+            if (task != nullptr)
+            {
+                task->run();
+                task_executed(task);
+            }
         }
     }
+    
+    /**
+    * Called by the thread pool after task has been executed.
+    * @param task Task which was executed.
+    */
+    void task_executed(abstract_task_container* task) override
+    {
+        std::lock_guard<std::mutex> lock(m_mutex_thread_pool);
+        m_tasks_to_process.push(task);
+        m_wait_for_task_executed.notify_one();
+    }
 
 protected:
     // Stores all task_containers with their position as an implicit id
     std::vector<std::unique_ptr<abstract_task_container>> m_tasks;
-    std::vector<std::thread> m_threads;
-    std::mutex m_mutex;
+    std::vector<abstract_task_container*> m_initial_tasks;
+    yats::thread_pool m_thread_pool;
+    std::mutex m_mutex_thread_pool;
+    std::condition_variable m_wait_for_task_executed;
+    std::queue<abstract_task_container*> m_tasks_to_process;
+
+    std::condition_variable m_wait_for_main_thread_task;
+    std::mutex m_mutex_main_thread;
+    std::queue<abstract_task_container*> m_tasks_to_process_in_main_thread;
+
+    std::thread m_task_distribution_thread;
+    std::condition_variable m_wait_for_run;
+    std::mutex m_mutex_run;
+    bool m_may_run;
+    bool m_terminate;
+    
+    std::size_t m_scheduled_task_count;
+    std::size_t m_tasks_processed{};
```

Comment:
You have way to many variables here. When you are handling the constroller thread in the threadpool, you save the variables that are requiered to control the lifetime of the thread because you can reuse the existing ones.